# Tutorial: The `.score()` Method: R-squared (R²) in Scikit-learn

Di subchapter sebelumnya, kita belajar metrik seperti MAE dan RMSE. Metrik-metrik ini sangat bagus untuk memberi tahu kita **besarnya kesalahan absolut** dalam satuan asli (misalnya, "kesalahan kita adalah 2.01 unit penjualan").

Namun, seringkali kita juga ingin menjawab pertanyaan yang berbeda:

> **"Seberapa baik model kita dibandingkan dengan tebakan yang paling sederhana sekalipun?"**

Di sinilah **R-squared (R²)** atau **Koefisien Determinasi** berperan. R² adalah metrik **relatif** yang mengukur seberapa banyak variasi dalam data target yang dapat dijelaskan oleh model kita.

---
### 1. Tujuan Pembelajaran

Di akhir notebook ini, Anda akan dapat:

* Menjelaskan intuisi di balik R-squared (R²) sebagai ukuran performa relatif.
* Mengimplementasikan perhitungan R² menggunakan metode `.score()` bawaan dari model Scikit-learn.
* Mengimplementasikan perhitungan R² menggunakan fungsi `r2_score` dari `sklearn.metrics`.
* Menginterpretasikan nilai R² dalam konteks bisnis yang praktis.


---
### 2. Setup: Menggunakan Model yang Sudah Ada

Kita akan melanjutkan dari titik terakhir kita di subchapter sebelumnya. Kita sudah memiliki model yang dilatih pada dataset "Advertising" dan prediksi yang sudah dibuat.


In [3]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

# Memuat dataset
data = {
    'TV': [230.1, 44.5, 17.2, 151.5, 180.8, 8.7, 57.5, 120.2, 8.6, 199.8, 66.1, 214.7, 23.8, 97.5, 204.1, 195.4, 67.8, 281.4, 69.2, 147.3],
    'Radio': [37.8, 39.3, 45.9, 41.3, 10.8, 48.9, 32.8, 19.6, 2.1, 2.6, 5.8, 24.0, 35.1, 7.6, 32.9, 47.7, 36.6, 39.6, 20.5, 23.9],
    'Newspaper': [69.2, 45.1, 69.3, 58.5, 58.4, 75.0, 23.5, 11.6, 1.0, 21.2, 24.2, 4.0, 65.9, 7.2, 46.0, 52.9, 114.0, 55.8, 18.3, 19.1],
    'Sales': [22.1, 10.4, 12.0, 16.6, 17.9, 7.2, 11.8, 13.2, 4.8, 15.6, 12.6, 17.4, 9.2, 9.7, 19.0, 22.4, 12.5, 24.4, 11.3, 14.6]
}
df = pd.DataFrame(data)

# Persiapan data dan model (sama seperti sebelumnya)
X = df[['TV', 'Radio', 'Newspaper']]
Y= df['Sales']

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=42)

model = LinearRegression()
model.fit(X_train, Y_train)

Y_pred = model.predict(X_test)

print("Setup selesai.")

Setup selesai.


---
### 3. Intuisi di Balik R-squared (R²)

R² membandingkan model kita dengan sebuah **model dasar (baseline model)** yang sangat naif. Model dasar ini tidak melakukan pembelajaran apa pun; ia hanya selalu menebak **nilai rata-rata** dari target (`y`).

**R² menjawab:** "Seberapa jauh lebih baik model kita dalam mengurangi kesalahan dibandingkan dengan hanya menebak rata-rata?"

* Jika model kita **sempurna**, kesalahannya nol, dan **R² = 1.0**.
* Jika model kita **tidak lebih baik** dari menebak rata-rata, **R² = 0.0**.
* Jika model kita **lebih buruk** dari menebak rata-rata, **R² bisa menjadi negatif**.


---
### 4. Implementasi dengan Metode `.score()`

Cara termudah untuk mendapatkan skor R² di Scikit-learn adalah dengan menggunakan metode `.score()` yang sudah ada di setiap model.

Metode ini secara otomatis membuat prediksi pada data yang Anda berikan (`X_test`) dan membandingkannya dengan jawaban sebenarnya (`y_test`) untuk menghitung R².


In [8]:
# menggunakan metode .score()

r2_score_method = model.score(X_test, Y_test)

print(f'r2_score_method = {r2_score_method:.4f}')

r2_score_method = 0.8789


---
### 5. Implementasi dengan `sklearn.metrics.r2_score`

Untuk konsistensi dan pemahaman yang lebih dalam, kita juga bisa menghitung R² menggunakan fungsi spesifik dari modul `sklearn.metrics`. Ini berguna jika Anda sudah memiliki nilai prediksi (`y_pred`) dan ingin menghitung skornya secara manual.

Polanya sama dengan metrik lain: `r2_score(y_true, y_pred)`.

In [7]:
from sklearn.metrics import r2_score

r2_score_function = r2_score(Y_test, Y_pred)

print(f'Skor R-squared dari metode r2_score = {r2_score_function:.4f}')

Skor R-squared dari metode r2_score = 0.8789


Seperti yang Anda lihat, hasilnya **sama persis**. Metode `.score()` hanyalah sebuah jalan pintas yang nyaman.

---
### 6. Interpretasi Hasil R²

Kita mendapatkan skor R² sekitar **0.8789**. Apa artinya ini?

> **"Sekitar 87.89% dari variasi atau sebaran dalam data penjualan (Sales) pada set pengujian dapat dijelaskan oleh model kita menggunakan fitur TV, Radio, dan Newspaper."**

Ini adalah skor yang sangat baik. Ini menunjukkan bahwa fitur-fitur yang kita pilih memiliki kekuatan prediktif yang sangat tinggi dan model kita berhasil menangkap hubungan tersebut. Sisa ~12.11% variasi mungkin disebabkan oleh faktor-faktor lain yang tidak ada dalam data kita (misalnya, aktivitas kompetitor, musim, dll.) atau *noise* acak.

**Peringatan:** R² yang tinggi tidak selalu berarti modelnya sempurna. Sebuah model bisa *overfitting* (terlalu menghafal data pelatihan) dan mendapatkan R² yang tinggi pada data pelatihan, tetapi buruk pada data pengujian. Itulah mengapa kita **selalu** melaporkan R² dari **data pengujian**.


---
### 7. Ringkasan

* Metrik seperti **MAE dan RMSE** memberi tahu kita **besarnya kesalahan** dalam satuan asli. Mereka menjawab, "Seberapa besar kesalahan kita?".
* **R-squared (R²)** memberi tahu kita **kekuatan penjelasan** model dalam persentase. Ia menjawab, "Seberapa baik model kita?".

Keduanya sangat penting dan saling melengkapi. Dalam laporan, sangat umum untuk menyajikan keduanya, misalnya: "Model kami dapat memprediksi penjualan dengan kesalahan tipikal (RMSE) sebesar 1.98 unit dan dapat menjelaskan 87.86% variasi dalam data (R²)."

**Selanjutnya:**
Kita sekarang memiliki perangkat lengkap untuk mengukur performa model. Di subchapter berikutnya, kita akan melihat bagaimana cara menganalisis kesalahan model secara **visual** menggunakan **Residual Plots**, yang dapat mengungkapkan masalah yang tidak terlihat dari metrik angka saja.